operator/internal/controller/pulumi/program_controller.go

```diff
@@ -0,0 +1,314 @@
+		if programIdentifier == "" {
+			http.Error(w, "Program identifier is required.", http.StatusBadRequest)
+			return
+		}
+
+		namespace, name, err := urlPathToProgramIdentifier(programIdentifier)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		proj, err := getProjectFile(context.Background(), p.k8sClient, namespace, name)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusNotFound)
+			return
+		}
+
+		// Marshal the project file to YAML.
+		data, err := yaml.Marshal(proj)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		// Create a tarball of the project file and write it to the response.
+		out, err := tarProgram(data)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		// Compress the tarball using gzip.
+		out, err = gzipFile(out)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		w.Header().Set("Content-Type", "application/gzip")
+		w.Header().Set("Content-Disposition", "attachment; filename=project.tar.gz")
+		w.Header().Set("Content-Length", fmt.Sprintf("%d", len(out)))
+		w.Write([]byte(out))
+	}
+}
+
+// getProjectFile retrieves a Program object from the Kubernetes API server and returns a valid Pulumi Yaml project file from it.
+func getProjectFile(ctx context.Context, kubeClient client.Client, namespace, name string) (*ProjectFile, error) {
+	program := new(pulumiv1.Program)
+	programKey := client.ObjectKey{
+		Name:      name,
+		Namespace: namespace,
+	}
+
+	err := kubeClient.Get(ctx, programKey, program)
+	if err != nil {
+		return nil, err
+	}
+
+	project := programToProject(program)
+
+	return project, nil
+}
+
+// programToProject wraps a Program object to a ProjectFile.
+func programToProject(program *pulumiv1.Program) *ProjectFile {
+	return &ProjectFile{
+		Name:        program.Name,
+		Runtime:     "yaml",
+		ProgramSpec: program.Program,
+
+		timeModified: program.Status.LastResyncTime,
+	}
+}
+
+//+kubebuilder:rbac:groups=pulumi.com,resources=programs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=pulumi.com,resources=programs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=pulumi.com,resources=programs/finalizers,verbs=update
+
+// Reconcile reconciles the Program object.
+func (r *ProgramReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.Info("Reconciling Program")
+
+	// Fetch the Program object.
+	program := new(pulumiv1.Program)
+	if err := r.Get(ctx, req.NamespacedName, program); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Update the status of the Program object to contain an updated URL.
+	program.Status.Artifact = &pulumiv1.Artifact{
+		URL: r.ProgramHandler.CreateProgramURL(program.Namespace, program.Name),
+	}
+	program.Status.LastResyncTime = metav1.Now()
```

Comment:
This line will make the object be more volatile than necessary.

Reply:
Removed, along with the `lastResyncTime` field.